zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/controller/FlowNodeInstanceQueryController.java

```diff
@@ -52,12 +58,12 @@ public ResponseEntity<FlowNodeInstanceSearchQueryResponse> searchFlownodeInstanc
   public ResponseEntity<FlowNodeInstanceItem> getByKey(
       @PathVariable("flowNodeInstanceKey") final Long flowNodeInstanceKey) {
     try {
-      return ResponseEntity.ok()
-          .body(
-              SearchQueryResponseMapper.toFlowNodeInstance(
-                  flownodeInstanceServices
-                      .withAuthentication(RequestMapper.getAuthentication())
-                      .getByKey(flowNodeInstanceKey)));
+      final FlowNodeInstanceEntity flowNode =
+          flownodeInstanceServices
+              .withAuthentication(RequestMapper.getAuthentication())
+              .getByKey(flowNodeInstanceKey);
+      final String name = xmlUtil.getFlowNodeName(flowNode);
+      return ResponseEntity.ok().body(SearchQueryResponseMapper.toFlowNodeInstance(flowNode, name));
```

Comment:
💭 We could move the name resolution into the service. That way, the entity the service layer returns contains the name, and the REST controller consumes this DTO. That makes the REST controller code a lot simpler (it mostly stays the same). The service layer is responsible for fetching the data correctly; the REST controllers shouldn't need to care. Any other consumer of the service layer would receive a "complete" entity result as well this way. Plus, this should be the task of the search layer, ultimately. For example, the RDBMS mapper might decide to deliver those values directly, no need for fetching it manually. This would be hidden in the search transformers then.

Reply:
This would make more sense. I'll give this a shot.

> For example, the RDBMS mapper might decide to deliver those values directly

I guess the service layer could check if the names are set already and then do nothing.

Reply:
The service layer shouldn't need to care either. The search layer must take care of delivering the right entity.

Reply:
But if the name resolution happens in the service then how can RDBMS decide to deliver them directly? Should we define a search client method for resolving the name?

Reply:
Ah, sorry, the name resolution should happen in the search layer, not the service layer 🙈 I started off with the idea of doing it in the service layer, then realized it should actually go into search 😄

Reply:
Yeah, that could work. I could define a search layer method with default `XmlUtil` implementation for now.

Reply:
But wait, we want to cache this in the next step and that's supposed to be in the REST API.
That can only work if the REST API calls the service for name resolution.

1. REST API get entity (service -> search)
2. REST API fetch & cache names (service -> search)

Reply:
And the search layer can't cache this? We can also sync on this, might be easier.

Reply:
There will be a follow-up ticket for moving the cache and name resolving into the search clients.